### ISSUE TITLE
Featured Experiment section is not correctly updated after following the flow to install both TxP and Notes

### ISSUE THREAD

**Issue by SoftVision-PaulOiegas** [mozilla/testpilot#3387]
**[Affected versions]:**
- Firefox Nightly 60.0a1 

**[Affected Platforms]:**
- Windows
- Mac OS
- Linux

**[Prerequisites]:**
- Following 5 prefs have been set or added on a new profile: "extensions.legacy.enabled = true" "xpinstall.signatures.required = false" "testpilot.env = dev" "extensions.webapi.testing = true" "extensions.install.requireBuiltInCerts = false".

**[Steps to reproduce]:**
1. Start Firefox browser with the profile from prerequisite and navigate to https://testpilot.dev.mozaws.net page.
2. Click on "Install test Pilot & Enable Notes" big green button on top.
3. Confirm the doorhanger and click on the "Add" button on the next one displayed.
4. Confirm again by clicking "OK" on the 3rd one.
5. Dismiss the Notes tour and observe the "Featured Experiment" section.

**[Expected result]:**
- The "Enabled" label appears on top left +  "View details" button and "Install test Pilot & Enable Notes" big green button are removed and replaced by "Manage" and "Feedback" ones.

**[Actual result]:**
- The area remains unchanged and "Install test Pilot & Enable Notes" big green button is still displayed.

**[Notes]:**
- If the button is clicked again, the "Featured Experiment" section is correctly updated and displays the right state.
- Refreshing the page, does not fixes the issue.
- This issue also occurs on stage server.
- Might be a race condition issue since sometimes we managed to have the section correctly updated.
- Here is a screen recording of the issue: https://goo.gl/TfSWWV.

**Comment by chuckharmston** [comment]
@fzzzy and I both spent some time yesterday attempting to reproduce this, but were unable to.

> Might be a race condition issue since sometimes we managed to have the section correctly updated.

@SoftVision-PaulOiegas Are there any additional STR you've been able to uncover to make this more reliably reproducible?

**Comment by SoftVision-PaulOiegas** [comment]
@chuckharmston @fzzzy We've tried to debug this further too and performed the next tests:
- We tried to do the steps both fast and slow in the idea of timing issue, but the same behavior happens no matter what.
- I tried with bandwidth throttling applications in the idea of reproducing this only on fast internet connections. No luck, 3G throttling behaves the same. 
- Cosmin tried with US VPN and the behavior was the same.
- We also tried to delete the temp folder where the xpi's are downloaded and reinstall both. This worked for a few tries but in the end it failed again.

However, we have observed that when this occurs you are left on https://testpilot.dev.mozaws.net/ page instead of https://testpilot.dev.mozaws.net/experiments/ and only at second click on the green button you get redirected to /experiments.

This made us realize that the issue does not reproduces if you navigate straight to https://testpilot.dev.mozaws.net/experiments/ page (which is possible for some reason) and choose to install TxP and enable Notes. 
Is there a chance you guys are trying to install from /experiments too, instead of the Homepage? That would explain why we are reproducing this.

**Comment by SoftVision-PaulOiegas** [comment]
I have excluded also our work network to be the problem. Used my Android phone to request the test Pilot desktop version and followed the Featured experiment flow. Same issue occurs even using 4G network.

So we are still betting on the testpilot.dev.mozaws.net/ vs testpilot.dev.mozaws.net/experiments installation. @chuckharmston Have you tried installing from the first also?

**Comment by marniepw** [comment]
@chuckharmston or @fzzzy any luck with this issue?

**Comment by SoftVision-PaulOiegas** [comment]
This behavior seems to be related with other issues too.
#2897 - Uninstalling Test Pilot add-on randomly leaves you on the same page instead of "Retire"
#3374 - [Featured Experiment] The "Enabled" label is no longer displayed if the experiment is enabled but Test Pilot was uninstalled

On both the issues, the uninstall flow fails and you are left hanging on /experiments instead on Homepage. This issue behaves somehow the same since installing both, leaves you on the Homepage, instead redirecting you to /experiments page.

**Comment by chuckharmston** [comment]
Update: I've still had no luck reproducing. Thanks for the extra info @SoftVision-PaulOiegas, I'll keep working on it.

**Comment by fzzzy** [comment]
I was able to repro once, but not again when I tried again immediately. One thing I noticed is that while I had started on / and expected to be redirected to /experiments after the tour, I was left on / once the repro happened.

There's some race condition with the addon being installed which sends a message to the webpage letting the webpage know the addon is installed.

**Comment by fzzzy** [comment]
Ok, I have a reliable repro now. When the tour pops up, quickly dismiss it with the "X" to close it instead of clicking through all the steps of the tour. Previously I would always go through all the tour steps, which takes longer and the addon is done installing by the time the user dismisses it.

**Comment by SoftVision-PaulOiegas** [comment]
> 5. Dismiss the Notes tour and observe the "Featured Experiment" section.

Well, this is part of my steps. 😄  But I'm glad you guys finally caught it. As I've said, this seems to be related to other issues too (#2897 and #3374) and yes, seems to be a race condition at the root of all 3.

**Comment by SoftVision-PaulOiegas** [comment]
The fix seems to be working for normal flows. Now if you dismiss the tour, it redirects you to the /experiments page. 👍 

However, we found two cases outside of the normal flow, where this is still not working:
1. One edge case that implies spamming "Esc" key after clicking on doorhanger "Add" button. Which leaves you on Homepage with both add-ons installed. Managed once to get also a blank /experiments page, so this one is more of a race condition.

2. Refreshing the page while tour is displayed. While this is somehow ok since you are on homepage, the right behavior would be to redirect you to the /experiments page. This exact behavior happens if you try to navigate to homepage with Test Pilot already installed. You are automatically redirected to /experiments.

While the first is an edge case that I don't think we should be worried about (unless you think otherwise). We will be logging the second one to be fixed on a future build.
